### PR TITLE
WIP: TTL caches

### PR DESCRIPTION
**Pull request by lennerd**
* [x] Add `TTLCache` class
* [x] Add ClientSideConfigurationCache with TTL of an hour.
* [ ] Add TTL to CategoryCache
* [ ] Add TTL to LicenseCache and SourceCache

### CODE REVIEW

File: src/lib/cache/URLDataCache.js [hunk 1]
```diff
@@ -1,55 +1,58 @@
...
+import TTLCache, { type TTLCacheOptions } from './TTLCache';
 
 // Provides a WhatWG-fetch-like API to make HTTP requests.
 // Caches response promises and returns an old promise if one is existing for the same URL.
 
+type URL = string;
+
 export default class URLDataCache<T> {
-  cache: { [key: string]: Promise<T> } = {};
+  cache: TTLCache<URL, Promise<T>>;
```
Comment (opyh) [inline src/lib/cache/URLDataCache.js]: +1: Composition > Inheritance 👍🏻

File: src/lib/cache/URLDataCache.js [hunk 2]
```diff
@@ -1,55 +1,58 @@
...
 
-  fetch(url: string, resolve: (data: T) => void, reject: (response: any) => void) {
-    this.constructor.fetch(url, { cordova: true }).then((response: Response) => {
+  constructor(options?: TTLCacheOptions) {
+    this.cache = new TTLCache<URL, Promise<T>>(options);
```
Comment (opyh) [inline src/lib/cache/URLDataCache.js]: 0: This could happen directly in the declaration above

File: src/lib/cache/URLDataCache.js [hunk 3]
```diff
@@ -1,55 +1,58 @@
...
-          resolve(fetchedData);
-        }, reject);
+        return this.constructor.getDataFromResponse(response);
       }
-      const error = new Error(response.statusText);
-      error.response = response;
-      return reject(response);
-    }, reject);
+
+      throw new new ResponseError(response.statusText, response)();
```
Comment (opyh) [inline src/lib/cache/URLDataCache.js]:
```suggestion
      throw new ResponseError(response.statusText, response);
```

File: src/lib/cache/URLDataCache.js [hunk 4]
```diff
@@ -1,55 +1,58 @@
...
 import { globalFetchManager } from '../FetchManager';
 import { t } from 'ttag';
+
 import ResponseError from '../ResponseError';
+import TTLCache, { type TTLCacheOptions } from './TTLCache';
 
 // Provides a WhatWG-fetch-like API to make HTTP requests.
 // Caches response promises and returns an old promise if one is existing for the same URL.
 
+type URL = string;
```
Comment (opyh) [inline src/lib/cache/URLDataCache.js]: -1: I think this is misleading, as URL looks like a complex type on first sight, and there is a `URL` class on window in the browser. I'd leave `string` as key type – it's clear what it does and it does not imply the thought that there might be some canonical URL recognition tech behind it. Also, right in the subclass it's no URL anymore, but a hostname.

File: src/lib/cache/TTLCache.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  constructor(options?: TTLCacheOptions) {
+    this.options = {
+      ttl: Infinity,
+      ...options,
+    };
+  }
+
+  set(key: K, value: V, ttl: number = this.options.ttl, now: number = Date.now()): boolean {
+    // Adding the value to the cache is not possible if ttl is zero.
+    if (ttl === 0) {
```
Comment (opyh) [inline src/lib/cache/TTLCache.js]: Would use `<= 0` here to be more nice to the API user :)

File: src/lib/cache/TTLCache.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  options: TTLCacheOptions;
+
+  constructor(options?: TTLCacheOptions) {
+    this.options = {
+      ttl: Infinity,
+      ...options,
+    };
+  }
+
+  set(key: K, value: V, ttl: number = this.options.ttl, now: number = Date.now()): boolean {
```
Comment (opyh) [inline src/lib/cache/TTLCache.js]: +1: I like that you can use this functionally without time as a side effect.

File: src/lib/cache/TTLCache.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    this.cache.set(key, item);
+
+    return true;
+  }
+
+  get(key: K, remove: boolean = false, now?: number): ?V {
+    const value = this.touch(key, now);
+
+    if (value == null) {
```
Comment (opyh) [inline src/lib/cache/TTLCache.js]: How about `===`?

File: src/lib/cache/TTLCache.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  }
+
+  remove(key: K): boolean {
+    return this.cache.delete(key);
+  }
+
+  touch(key: K, now: number = Date.now()): ?V {
+    const item = this.cache.get(key);
+
+    if (item == null) {
```
Comment (opyh) [inline src/lib/cache/TTLCache.js]: How about `===`?

File: src/lib/cache/TTLCache.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      this.remove(key);
+
+      return null;
+    }
+
+    return item.value;
+  }
+
+  touchAll(now?: number): void {
+    for (const key of this.cache.keys()) {
```
Comment (opyh) [inline src/lib/cache/TTLCache.js]: How about this?

```suggestion
this.cache.keys().forEach((key) =>  this.touch(key, now));
```

…as this does not use the regenerator runtime

File: src/lib/cache/URLDataCache.js (hunk 2 above)
Comment (lennerd) [reply src/lib/cache/URLDataCache.js]: How do you pass options into the cache then?

File: src/lib/cache/URLDataCache.js (hunk 3 above)
Comment (lennerd) [reply src/lib/cache/URLDataCache.js]: Good catch!

File: src/lib/cache/TTLCache.js (hunk 7 above)
Comment (lennerd) [reply src/lib/cache/TTLCache.js]: As explained in slack, this also tests against both `null` and `undefined`.

File: src/lib/cache/TTLCache.js (hunk 9 above)
Comment (lennerd) [reply src/lib/cache/TTLCache.js]: I thought the regenerator runtime is only included when used with Generator or async functions? The generator class itself is not needing the polyfill AFAIK. Will investigate.

Also the regenerator runtime is already included because of the async functions when fetching intial props.

Comment (opyh) [reply src/lib/cache/TTLCache.js]: https://github.com/airbnb/javascript/issues/1271 has a long discussion of many aspects of this if you're interested :)

File: src/lib/cache/TTLCache.js (hunk 7 above)
Comment (opyh) [reply src/lib/cache/TTLCache.js]: Ah – if this was the intention, a small comment would be cool :)

File: src/lib/cache/TTLCache.js (hunk 9 above)
Comment (opyh) [reply src/lib/cache/TTLCache.js]: (Also, you seem to be right, and to have that for loop there is totally okay.)

File: src/lib/cache/TTLCache.js [hunk 10]
```diff
@@ -33,11 +43,23 @@ class TTLCache<K, V> {
...
+      this.shift();
+    }
+
     this.cache.set(key, item);
 
     return true;
   }
 
+  shift() {
+    const key = this.cache.keys().next().value;
```
Comment (opyh) [inline src/lib/cache/TTLCache.js]: I'm not sure – do have ES6 maps have a specified order? If not, this could remove the last inserted item!

Comment (lennerd) [reply src/lib/cache/TTLCache.js]: Yes, they have: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#Objects_and_maps_compared

> The keys in Map are ordered while keys added to object are not. Thus, when iterating over it, a Map object returns keys in order of insertion.

Comment (opyh) [reply src/lib/cache/TTLCache.js]: Super :)

Comment (lennerd) [reply src/lib/cache/TTLCache.js]: I will add a comment, to clarify this for future readers. ;)

**Comment by lennerd** [comment]
Öhm … This is not finished nor ready to be merged.
